community/sustainability_effort.md

```diff
@@ -0,0 +1,80 @@
+# Argo Workflows Sustainability Effort
+
+Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable. 
+
+We are trying an Incentive System in which contributors pledge a certain number of hours per week (average) and in return have their PRs prioritized for review.
+
+Help is also needed for triaging new incoming bugs by prioritizing them with `P0`, `P1`, `P2`, and `P3` labels.
+
+## Commitment
+
+The Argo project currently has 4 designated [roles](https://github.com/argoproj/argoproj/blob/main/community/membership.md):
+- Member
+- Reviewer
+- Approver
+- Lead
+
+For those who focus on just one area of the code (such as UI), the `Reviewer` and `Approver` roles can optionally be scoped to just that area.
+
+Any person in a designated role (or even those who are not yet Members) can sign up to participate in this Incentive System.
+Participants are expected to try to advance in roles.
+There are different expectations depending on the role:
+- Member plus those not yet Member: should focus primarily on authoring PRs, should average a minimum of 8 hours per week; can also assist with triaging bugs or optionally reviewing PRs
+- Reviewer: should author PRs and review PRs to move into the "Approver" role, should average a minimum of 2 hours per week of either PR review time or triaging bugs
```

Comment:
```suggestion
- Reviewer: should author and review PRs, averaging a minimum of 2 hours per week of either PR review time or triaging bugs
```

- same structure as above
- It already says above that "participants are expected to try to advance in roles", so the Approver portion is redundant

Reply:
To clarify, does this mean that a Reviewer+'s authorship time is not considered as part of the system, but only review and triage time is conisdered?

Reply:
Also what about responding to Slack and GitHub Discussions? I spend a lot of time in those (although discussion volume has decreased recently while issues have increased) and those are considered for Membership requests.

Reply:
As for the first comment, I specifically added 'to move into the "Approver" role' as a clarification for why it was distinct from what "Approver" was doing. I think it could be useful, no?

Reply:
The Slack and Github Discussion thing is interesting - perhaps we should expand into that as well.

Reply:
> As for the first comment, I specifically added 'to move into the "Approver" role' as a clarification for why it was distinct from what "Approver" was doing. I think it could be useful, no?

Since progressing through roles is the baseline expectation, that sounds redundant to me.

In terms of distinctions, I suppose _within_ the incentive system itself (this specific doc), the _expectations_ (this specific list) are not necessarily altogether different for Reviewers vs Approver+. The Membership [guidelines](https://github.com/argoproj/argoproj/blob/main/community/membership.md) do list some other expectations, if you wanted to include some of those here, but I'm not sure if they're necessarily relevant to the incentive system _itself_, which is currently just about prioritizing PRs and so code review is the biggest piece.

Also what are your thoughts on considering PR authorship for Reviewer+? Might have missed that [comment above](https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1433402502) (in between the other two).
Both reviews and authorship are needed for maintenance, but we can certainly say that some activities (including Slack and GH Discussions) are out-of-scope of the incentive system.
To be specific, I don't necessarily have an opinion either way on non-review activities, just want to make sure we consider all possible contributions and whether they are in- or out-of-scope for this incentive system (we could also explicitly list what is out-of-scope as well)

Reply:
> Since progressing through roles is the baseline expectation, that sounds redundant to me.
Let's come back to this one since it's more trivial than the rest.

> In terms of distinctions, I suppose _within_ the incentive system itself (this specific doc), the _expectations_ (this specific list) are not necessarily altogether different for Reviewers vs Approver+. The Membership [guidelines](https://github.com/argoproj/argoproj/blob/main/community/membership.md) do list some other expectations, if you wanted to include some of those here, but I'm not sure if they're necessarily relevant to the incentive system _itself_, which is currently just about prioritizing PRs and so code review is the biggest piece.
> 
> Also what are your thoughts on considering PR authorship for Reviewer+? Might have missed that [comment above](https://github.com/argoproj/argo-workflows/pull/12359#discussion_r1433402502) (in between the other two). Both reviews and authorship are needed for maintenance, but we can certainly say that some activities (including Slack and GH Discussions) are out-of-scope of the incentive system. To be specific, I don't necessarily have an opinion either way on non-review activities, just want to make sure we consider all possible contributions and whether they are in- or out-of-scope for this incentive system (we could also explicitly list what is out-of-scope as well). It is, after all, entirely optional, so not all types of contributions have to necessarily be included

As far as PR authorship for Approvers+: since this person has arrived at "Approver" level, I think that's sufficient, and don't have any specific expectation for authoring PRs. As far as PR authorship for Reviewers: yes, I do see some inconsistency between what I indicated for "Reviewer" and what I indicated for "Member-". On the other hand, I didn't want to ask for too much, and wanted to focus on what I considered to be important. So, I think I'd like to maybe just keep that as is.? 

As far as the Github Discussions and Slack Discussions, I guess we are seeing that unfortunately there's a mountain of work to be done. :( Anything we think is important for the health of the repo should probably be included. For slack I assume we are referring to `argo-workflows` channel and `argo-wf-contributors` channel? And I assume that generally speaking it would be Reviewer+ responding on `argo-wf-contributors` channel, but anyone responding on `argo-workflows` channel and Github Discussions?

Perhaps I can include all of these activities then, but make it so that if you're a "Reviewer+", the most important thing is that we address the "prioritized" PRs, making sure that those have an assignee.

Speaking of "prioritized" label: maybe will rename that to "prioritized-pr" so nobody accidentally applies it to an Issue? Or if you have a preference of something else, let me know.

Reply:
I have made all of these changes now (except for renaming the label).

Reply:
> So, I think I'd like to maybe just keep that as is.?

I don't really have much of an opinion, so personally, I think it's fine either way.

> As far as the Github Discussions and Slack Discussions, I guess we are seeing that unfortunately there's a mountain of work to be done. :(

Yea I find support to be one of the biggest & oft unexpected parts of maintaining something.
_Personally_, in my own repos, I disable Discussions and have no messaging forum. The docs should be primary self-help resource and the focus should be on improvements to the docs rather than on support. Discussions can be a bit distracting, as you'll get support questions in both Issues and Discussions, except only one of them has issue templates etc 😕 
But CNCF strongly prefers having both.
And also support questions will come in one way or the other, it's more of a question of where. Issues are much more accessible to search than Slack or Discussions, so if you're gonna get support questions, might as well put them there. If GH had a more nuanced implementation of Discussions and better tooling for them, that would be ideal but 😕 

That's just my opinion / thought process, for reference. Argo certainly gets more support questions as it's fairly complicated with lots of configurability and also requires a k8s cluster (and also k8s familiarity) to get started, so folks tend to have more upfront questions as a result. That's my observation so far at least.

> Anything we think is important for the health of the repo should probably be included.

I do think support is important, but I also don't like users relying on free support and the dynamic that creates -- i.e. high expectations of free volunteer help.
Vendors with support contracts I think should be more of the norm, but as a whole it is part of one of the current debates in OSS that does not really have a good solution or standard in place. Users just getting started or exploring won't want to make a support contract just yet; although PoCs & pilots also exist for that purpose in the sales process.

The reason I spend more time on those was because:
- initially, knowledge as a user is how you can contribute best
  - linking to docs and contributing to docs after answering questions was another area I focused on and that I think is good for initial contributions
- it's a lot easier to participate there, especially from my phone, than on PRs and code reviews etc.

If we include those activities, then I can add my name as I already spend a sizeable amount of time there even during months I'm less active

> For slack I assume we are referring to `argo-workflows` channel and `argo-wf-contributors` channel?

Yes mostly the former

> And I assume that generally speaking it would be Reviewer+ responding on `argo-wf-contributors` channel, but anyone responding on `argo-workflows` channel and Github Discussions?

More or less yes. I think Member+ or even interested contributors would fit `#argo-wf-contributors`. I certainly used it a good bit as a Member and we do have contributors without roles there too.

> Perhaps I can include all of these activities then, but make it so that if you're a "Reviewer+", the most important thing is that we address the "prioritized" PRs, making sure that those have an assignee.

If we do include those, then yes I'd agree (again I don't really have a strong opinion on including them or not).
I think that makes the most sense to ensure the main target is still encouraging PR review, where we need the most help.

> Speaking of "prioritized" label: maybe will rename that to "prioritized-pr" so nobody accidentally applies it to an Issue? Or if you have a preference of something else, let me know.

Agree.
My ideal might be `triaged/prioritized-pr` or some category; personally I'd rename all `P0`-`P3` into the same `triaged` category as well (`triaged/P3` etc). A different category name could be `priority/` or `meta/` or something as well. `P3` etc is not specified to be for bugs in its name either.
I use categories pretty heavily in [other repos I maintain](https://github.com/ezolenko/rollup-plugin-typescript2/labels) as most things can be organized across a few categories. Tbh primary reason I'd want Approver status at this point is so that I can edit issue titles (which helps searchability and accuracy), labels, etc on my own 😅 

But nbd either way at this point of what to name it and whether to have a category or not. Not super important to this doc how its named